lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,472 @@
+//===- CheckCombCycles.cpp - FIRRTL check combinational cycles --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the FIRRTL combintational cycles detection pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SCCIterator.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+class CombNodeImpl;
+}
+
+using CombNode = CombNodeImpl *;
+using ValueOrOpType = PointerUnion<Value, Operation *>;
+
+//===----------------------------------------------------------------------===//
+// CombNodeImpl class
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// This is the implementation of a combinational node.
+class CombNodeImpl {
+public:
+  CombNodeImpl(ValueOrOpType valOrOp, unsigned fieldID = 0)
+      : valOrOp(valOrOp), fieldID(fieldID) {}
+
+  bool operator==(const CombNodeImpl &other) const {
+    return valOrOp == other.valOrOp && fieldID == other.fieldID;
+  }
+
+  CombNode *child_begin() { return children.begin(); }
+  CombNode *child_end() { return children.end(); }
+
+  void addChild(CombNode child) { children.emplace_back(child); }
+  void clearChilds() { children.clear(); }
+
+  ValueOrOpType getValOrOp() const { return valOrOp; }
+  unsigned getFieldID() const { return fieldID; }
+
+  llvm::hash_code hash_value() const {
+    return llvm::hash_combine(valOrOp.getOpaqueValue(), fieldID);
+  }
+
+private:
+  /// The underneath value or operation of this node.
+  ValueOrOpType valOrOp;
+
+  /// The field ID this node is referencing to.
+  unsigned fieldID;
+
+  /// Store all children of this node in the graph.
+  SmallVector<CombNode, 4> children;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// CombGraph class
+//===----------------------------------------------------------------------===//
+
+namespace {
+// This is a combinational graph contains a set of nodes.
+class CombGraph {
+public:
+  /// Get a node from the graph if it exists. Otherwise, allocate a new node and
+  /// add to the graph.
+  CombNode getOrAddNode(ValueOrOpType valOrOp, unsigned fieldID = 0) {
```

Comment:
This seems like it is effectively replicating the SSA graph into another datastructure.  This could make sense for SFC because it doesn't have SSA, can we avoid this in MFC?

Reply:
This is a good point. I tried to work on the SSA graph and use a customized iterator to iterate on a value/operation's children. But I found it's hard to do so, because directed edges in the combinational graph don't always flow from operands to results. For example:

1. `ConnectOp` connects the `src` operand to the `dest` operand;
2. For `InstanceOp`, we need to query `CombPaths` to determine which pairs of IOs are connected combinationally;
3. For `MemOp`, we need to connect the `en/addr` subfield and `(r)data` subfield if the read latency is zero;

And for some of them, it's not trivial to get the directed edges, such as `MemOp`. Therefore, I decided to construct a new graph for this :D

Reply:
Do the graph traits take references to the nodes, or do they copy them around? If it's the latter, you might be able to create the `CombNode` as a temporary thing on the stack that just lives inside the graph traits iterator. Then you can dump all the custom handling of ops in there, but don't need to push things into the allocator?

I'm looking at a similar thing in InferWidths, where it would be nice to actually not create the `Expr` graph shadowing the IR, but rather run on the IR directly.